### ISSUE TITLE
Guidance on in_parallel usage

### ISSUE THREAD

**Issue by bcthomas** [lostisland/faraday#354]
TL;DR What's the best way to write a client library with Faraday/em-http?

I'm writing a grape-based API (active_record for my ORM support) behind a Goliath server as well as the associated client library.  The client code uses Faraday and is styled loosely on what I've seen in the Twitter client gem.  I have a Client.get and Client.post, both using Client.request etc.

I am using the em-http faraday adapter.  It took me a while to get my code to hit the server in parallel; first I had to become aware of the in_parallel method, then I was stuck on getting that working until I implemented an on_complete block in my get requests.  Now I have a working library, but I have noticed a few issues.

My question is this: is there a preferred way to implement a parallel-request client using faraday? I have searched the examples in the wiki, and Github open/closed issues (and SO), but so far only find rather simple examples (e.g. https://gist.github.com/igrigorik/802395) which I am having trouble mapping to my scenario.

The relevant code from my library is here:

``` ruby
# from Client (client.rb)
      # set up the main Faraday loop
      #
      # @return [Faraday::RackBuilder]
      def middleware
        @middleware ||= Faraday::RackBuilder.new do |faraday|
          faraday.request :url_encoded
          # Handle error responses
          faraday.response :raise_error
          # Parse JSON response bodies
          faraday.response :parse_json
          # Set default HTTP adapter
          faraday.adapter :em_http
          #faraday.adapter :em_synchrony
        end
      end

      # connection options for faraday
      #
      # @return [Hash]
      def connection_options
        {
          :builder => middleware,
          :headers => {
            :accept => 'application/json',
            :user_agent => user_agent,
          },
          :request => {
            :open_timeout => 36,
            :timeout => 300,
          }
        }
      end

      # Returns a Faraday::Connection object
      #
      # @return [Faraday::Connection]
      def connection
        @connection ||= Faraday.new("#{configuration.base_uri}/#{configuration.api_version}", connection_options)
      end

# from Resource (resource.rb)
  module Resource
    # parallel request
    #
    # @param [Array] array of ids to fetch
    def find_in_parallel(resource_ids = [])
      results = []
        Client.connection.in_parallel do
          resource_ids.each do |rid|
            Client.get("#{resource_path}/#{rid}").on_complete do |env|
              results << env
            end
          end
        end
        $stderr.puts "results is #{results.size}"
      end

      results.collect! { |r| new(r.body) }
    end

    # generic resource request
    #
    # @params [Hash, Fixnum, Array] search params or ids
    def find(options = {})
      if options.is_a? Fixnum
        new(Client.get("#{resource_path}/#{options}").body)
      elsif options.is_a? Hash
        new(Client.get(resource_path, options).body)
      elsif options.is_a? Array
        find_in_parallel(options)
      end
    end
end
```

The library is meant to be called like this...

``` ruby
Client.configure do |config|
  config.api_key     = 'test'
  config.api_secret  = 'secret'
end
ResourceExample.find(42) # retrieve a single element
ResourceExample.find([1,2,3,4]).each { |x| ... } # retrieve, in parallel, many elements
```

The first problem, the above code in `find_in_parallel` works fine, UNLESS there are a huge number of requests (which is the nature of my data: parents can have many thousands of children etc.).  For example, 100, 200, 500 are no problem, but after i get up above 1800+ requests, the library fails with this...

```
.gem/ruby/2.0.0/gems/faraday-0.9.0/lib/faraday/adapter/em_http.rb:208:in
`run': unable to resolve server address (Faraday::ClientError)
```

In this specific test, 1015 requests hit the server and succeeded.  817 failed with this error.  (note from the adapter configuration above, I also tried em-synchrony, but this adapter failed too - the only difference is it didn't report back the 817 failed requests).  Splitting the requests up (e.g. 4 X 500 requests) works, so I changed my code to this...

``` ruby
    MAX_PARALLEL = 200
    def find_in_parallel(resource_ids = [])
      results = []
      resource_ids.each_slice(MAX_PARALLEL) do |slice|
        Client.connection.in_parallel do
          slice.each do |rid|
            Client.get("#{resource_path}/#{rid}").on_complete do |env|
              results << env
            end
          end
        end
        $stderr.puts "results is #{results.size}"
      end

      results.collect! { |r| new(r.body) }
    end
```

As you can see, I am just splitting up the in_parallel requests to a preset max.  I can live with that, but this is what I'd like advice on - this way seems hackish and incorrect.  I am planning on implementing a paging system into the library (as well as a cache), but I wanted to understand what's going here first.

Any comments?  Is my thinking/code in find_in_parallel off?  How is it that I'm getting this specific error? It seems like after a certain number or requests, the client just stops sending anything to the server.  Any comments or guidance appreciated, and thanks.
